lib/booking_manager.rb

```diff
@@ -0,0 +1,184 @@
+require_relative 'room'
+require_relative 'reservation'
+# require 'pry'
+
+module Hotel
+  class BookingManager
+    attr_accessor :rooms, :reservations, :room_calendar
+
+    def initialize(number_rooms)
+      @rooms = populate_room_list(number_rooms) #(20)
+      @reservations = make_reservation_list
+      @room_calendar = make_room_calendar(number_rooms)
+    end
+
+    # Create list of rooms as list of room Instances
+    def populate_room_list(number_rooms)
+      rooms = []
+      num = 1
+
+      number_rooms.times do |room|
+        room = Room.new(num)
+        #
+        rooms << room
+        num += 1
+      end
+
+      return rooms
+    end
+
+
+    # Create array to store all of reservations
+    def make_reservation_list
+      reservations = []
+      return reservations
+    end
+
+
+    # Create list of rooms with reserved dates.
+    def make_room_calendar(number)
+      @room_calendar = {}
+
+      @rooms.each do |room|
+        dates_reserved = {}
+        @room_calendar[room] = dates_reserved
+      end
+
+      return @room_calendar
+    end
+
+
+    # Method to add a reservation to list of reservations
+    def add_reservation(reservation)
+      @reservations << reservation
+    end
+
+
+    # Add reservation date range to hash of room reserved dates
+    def add_reservation_to_calendar(reservation)
+      date = reservation.start_date
+
+      reservation.number_nights.times do
+        @room_calendar[reservation.room][date] = reservation # dependency
+        date += 1
+      end
```

Comment:
I like the idea of using a hash of hashes to keep track of the reservations for each date for each room. This definitely works to solve the problem.

However, this ends up being an example of tight coupling. The information about what dates a reservation is for is repeated here in the `BookingManager`. That means that if a reservation's dates ever change, the program needs to know to come back here to the `BookingManager` and update this list - otherwise the logic to reserve a room will be broken. From the outside it's not obvious this should be true.

The way to resolve this would be to make the reservations the final source of truth about what dates they're for. When the `BookingManager` needs to build a list of reservations for a certain date, or figure out which rooms are available, it would have to loop through the list of all reservations to do the work, probably calling some method like `Reservation#overlap?(start_date, end_date)`. This may feel like more effort on your part, but in the long run it would make this design less fragile.

Reply:
Are you saying instead of having BookingManager have a calendar of Rooms with Reservations, that that information should only be accessible by ..one instance variable?Where else would the reservation date be changed besides in the add_reservation/add_reservation_to calendar method? (I have now combined those two. And maybe only need one.) Each reservation does have info on what date it is for.